File: matlab/srht.m
function [PHx] = srht(x, p, ordering)
% SRHT  Our implementation of the fast subsampled/trimmed Walsh-Hadamard transform
%
% We use the approach outlined in "Fast Dimension Reduction..." by Ailon + Liberty
% There are additionaly details in "Unified Matrix Treatment of the Fast Walsh-Hadamard Transform" by Fino + ...
%
% Inputs: x        Input vector (or matrix) of size [2^m k] for some m
%         p        Subsampling vector of size [2^m 1]
%         ordering 'hadamard'/'natural' or 'sequency' (<- not implemented)
%
% Outputs: PHx    subsampled/trimmed Hadamard transform of x using the order
%                 prescribed by ordering
%
% Note: This smart implementation of the SRHT isn't always faster than the naive
%       implementation of the SRHT using our fht.m:
%           Hx = fht(x, 'hadamard');
%           PHx = Hx(p,:);
%
% Alg: "Fast Dimension Reduction ..." Ailon + Liberty, 2010
%
% Deps: signal/bitrevorder?

%TODO handle other orderings (e.g. sequency)

if nargin < 3
   ordering = 'hadamard';
end

[m,n] = size(x);
[bool, e] = ispow2(m);
if ~bool && e >= 1
   error('srht: Invalid transform length.  Transform dimensions should be 2^e, e >= 1.');
end

[inds, I] = sort(p(:), 1, 'ascend'); % sort p here for faster searching later
Iinv(I) = 1:numel(I); % inverse permutation

PHx = srht_rec(x, e, inds, ordering);
PHx = PHx(Iinv,:) / sqrt(2)^e; % normalization and revert to user's subsampling order

end

function [PHx] = srht_rec(x, e, p, ordering)
% SRHT_REC  Recursive implementation of SRHT
%
% Note that this doesn't multiply by any 1/sqrt(2) factors
%
% Alg: "Fast Dimension Reduction ..." Ailon + Liberty, 2010

   switch ordering
   case {'hadamard', 'natural'}
      if e == 0
         error('srht_rec: Invalid transform length.  Transform dimensions should be 2^e, e >= 1.');
      
      % this case is small enough to compute Hx and then subsample
      elseif e == 1
         Hx = [1 1; 1 -1]*x;
         PHx = Hx(p,:);
      
      % this case is small enough to compute Hx and then subsample
      elseif e ==2
         Hx = [1 1 1 1; 1 -1 1 -1; 1 1 -1 -1; 1 -1 -1 1]*x;
         PHx = Hx(p,:);
      
      else
         half = pow2(e-1);
         [p1, p2, m] = split(p, half);
         k1 = numel(p1); k2 = numel(p2);

         x1 = x(1:half,:); % don't know if recursion will pass by reference or copy... prolly by ref
         x2 = x(half+1:2*half,:);
         
         % modify p2 so indexing on next level down matches up with length of x1, x2
         p2_mod = p2 - half;

         PHx = zeros(k1+k2,size(x,2));
         if k1 > 0
            PHx(1:k1,:) = srht_rec(x1+x2, e-1, p1, ordering);
         end

         if k2 > 0
            PHx(k1+1:k1+k2,:) = srht_rec(x1-x2, e-1, p2_mod, ordering);
         end
      end

   case {'sequency'}
      %TODO? see eqution (11) of Fino + Algazi (1976).
      error('srht_rec: not implemented');
   
   otherwise
      error('srht_rec: unsupported ordering: %s', ordering);
   end

end

function [p1, p2, m] = split(p, val)
% SPLIT  split a vector (sorted) along first column at value val via binary search
%
% Inputs: p    sorted vector of unique values
%         val  the value where we want to split
%
% Outputs: m   split index
%          p1  p up through val
%          p2  the rest of p
%
% Note: Due to the convenient indexing behavior in MATLAB, we sometimes will
%       index something as p(1:0), which returns an empty [1 0] array.  If
%       you do size(p(1:0),1) of this [1 0] array, you'll see size 1, even 
%       though the array is empty.  You can use numel(p(1:0)) == 0 or use
%       isempty to determine the array is empty.
%
% Alg: https://en.wikipedia.org/wiki/Binary_search_algorithm

L = 1; R = numel(p);

while true
   if L > R
      % binary search failed
      % assuming p is actually sorted, val may not exist in p
      % we'll go ahead and split at m, which should still split p appropriately
      m = R;
      p1 = p(1:R);
      p2 = p(R+1:end);
      return 
   end

   m = floor((L+R)/2);

   if p(m) < val
      L = m+1;

   elseif p(m) > val
      R = m-1;

   else % p(m) == val, split at m and we're done
      p1 = p(1:m);
      p2 = p(m+1:end);
      return
   end
end

end

File: julia/srht.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <math.h>

// use SIMD? (be sure to compile and link appropriately)
#define USE_SIMD 0
#if USE_SIMD
   #include <x86intrin.h> 
#endif

void split(unsigned *p,  unsigned len_p, unsigned val,
         unsigned *mp, unsigned **p1, unsigned *k1, unsigned **p2, unsigned *k2);


void srht_rec_had(double *pPHx, double *x, unsigned m, unsigned n, unsigned L,
         unsigned *pinds, unsigned len_p, unsigned len_inds_full, unsigned pstart)
{
// arrays coming from julia are column-major (Fortran) ordered

   static unsigned j,i;

   if ( L >= 3 ) {
      unsigned half, jhalf;
      unsigned mp, *p1, k1, *p2, k2;

      half = 2<<(L-2); // 2^(L-1)

      split(pinds, len_p, half-1, &mp, &p1, &k1, &p2, &k2);

      //printf("k1 = %d, k2 = %d\n", k1, k2);

      double *tmp = malloc(half*n*sizeof(double));
      if ( tmp == NULL ) {
         printf("srht.c: memory error.\n"); exit(-1);
      }
      
      //TODO (inline) function?
      #if(USE_SIMD) 
      // tmp = x1 + x2 with SIMD
      // half is always a factor of 2 (or 4), so no cleanup
      printf("SIMD NOT IMPLEMENTED\n");
      exit(-1);

      #else
      // tmp = x1 + x2 with a simple loop
      for ( j=0; j<n; ++j ) {
         jhalf = j*half;
         for ( i=0; i<half; ++i ) { 
            tmp[jhalf+i] = x[2*jhalf+i] + x[2*jhalf+i+half];
         }
      }
      #endif
      
      if ( k1 > 0 ) {
         srht_rec_had(pPHx, tmp, half, n, L-1, p1, k1, len_inds_full, pstart);
      }
      
      //TODO (inline) function?
      #if(USE_SIMD) 
      // tmp = x1 - x2 with SIMD
      // half is always a factor of 2 (or 4), so no cleanup
      printf("SIMD NOT IMPLEMENTED\n");
      exit(-1);

      #else
      // tmp = x1 - x2 with a simple loop
      for ( j=0; j<n; ++j ) {
         jhalf = j*half;
         for ( i=0; i<half; ++i ) { 
            tmp[jhalf+i] = x[2*jhalf+i] - x[2*jhalf+i+half];
         }
      }
      #endif
      
      if ( k2 > 0 ) {
         // modify p2 so indexing on next level down matches up with length of x1, x2
         for ( i=0; i<k2; ++i ) {
            p2[i] -= half;
         }

         srht_rec_had(pPHx, tmp, half, n, L-1, p2, k2, len_inds_full, pstart+k1);
      }

      free(p1);
      free(p2);
      free(tmp);
   }

   else if ( L == 2 ) {
      static double x0, x1, x2, x3; //TODO profile the static part
      static unsigned p_ptr, jm, jl;

      for ( j=0; j<n; ++j ) {
         jm = j*m;
         jl = j*len_inds_full;
         x0 = x[jm];
         x1 = x[jm+1];
         x2 = x[jm+2];
         x3 = x[jm+3];
         
         p_ptr = pstart;
         //printf("len_p = %d\n", len_p);
         for ( i=0; i < len_p; ++i ) { //TODO lots of decisions in this inner loop...
                                       //     Set x0,...,x3 equal to transform values, then
                                       //     play p_ptr games to add in the right one to the
                                       //     right spot?
            //printf("p_ptr = %d, len_p = %d\n", p_ptr, len_p);
            //printf("pinds[i] = %d\n", pinds[i]); 
            
            //printf("jl+p_ptr = %d\n", jl+p_ptr);

            if ( pinds[i] == 0 ) {
               pPHx[jl+p_ptr] = x0+x1+x2+x3;
               //printf("I'm here: 0\n");
               p_ptr += 1;
            }

            else if ( pinds[i] == 1 ) {
               pPHx[jl+p_ptr] = x0-x1+x2-x3;
               //printf("I'm here: 1\n");
               p_ptr += 1;
            }

            else if ( pinds[i] == 2 ) {
               pPHx[jl+p_ptr] = x0+x1-x2-x3;
               //printf("I'm here: 2\n");
               p_ptr += 1;
            }

            else if ( pinds[i] == 3 ) {
               pPHx[jl+p_ptr] = x0-x1-x2+x3;
               //printf("I'm here: 3\n");
               p_ptr += 1;
            }
         }
      }
   }

   else if ( L == 1 ) {
      printf("L = 1 NOT IMPLEMENTED\n");
      exit(-1);
   }

   else {
      printf("L went bad: L = %d\n", L);
      exit(-1);
   }

   return;
}


void split(unsigned *p,  unsigned len_p, unsigned val,
         unsigned *mp, unsigned **p1, unsigned *k1, unsigned **p2, unsigned *k2)
{
// Alg: https://en.wikipedia.org/wiki/Binary_search_algorithm

   unsigned L = 0, R = len_p-1, i;
      
   //printf("val = %d\n", val);
   //printf("len_p = %d\n", len_p);
   //for ( i=0; i<len_p; ++i ) {
   //   printf("p[i] = %d\n", p[i]);
   //}

   while ( 1 ) {
      if ( L > R ) {
         // binary search failed
         // assuming p is actually sorted, val may not exist in p
         // we'll go ahead and split at m, which should still split p appropriately
         *mp = R;
         //printf("bsearch failed\n");
         break;
      }
   
      *mp = (unsigned)(floor((double)(L+R)/2.0));
      //printf("L,R,*mp = %d, %d, %d\n", L,R,*mp);

      if ( p[*mp] < val ) {
         L = *mp + 1;
      }

      else if ( p[*mp] > val ) {
         if ( *mp == 0 ) break; // if len_p == 1, R can go negative (but unsigned)
         R = *mp - 1;

      }

      else { // p[*mp] == val
         break;
      }
   }
   //printf("L = %d, *mp = %d, R = %d\n", L, *mp, R);

   if ( *mp == 0 && p[*mp] > val ) *k1 = 0;
   else *k1 = *mp+1;

   if ( *mp == len_p-1 && p[*mp] < val) *k2 = 0;
   else *k2 = len_p-*k1;
   
   //*k1 = *mp+1;
   *p1 = malloc(*k1*sizeof(*p1));
   if ( p1 == NULL ) {
      printf("srht.c: memory error.\n"); exit(-1);
   }

   for (i=0; i<*k1; ++i) {
      (*p1)[i] = p[i]; // note that *p1[i] == *(p1[i]), as all postfix operators 
                       // take precedence over all prefix operators
      //printf("split: *p1[%d] = %d\n", i, (*p1)[i]);
   }

   //*k2 = len_p-*k1;
   *p2 = malloc(*k2*sizeof(*p2));
   if ( p2 == NULL ) {
      printf("srht.c: memory error.\n"); exit(-1);
   }

   for (i=0; i<*k2; ++i) {
      (*p2)[i] = p[i+*k1];
      //printf("split: *p2[%d] = %d\n", i, (*p2)[i]);
   }
}

//void srht_rec_had(double *pPHx, double *x, unsigned m, unsigned n, 
//         unsigned start, unsigned stop, unsigned L, unsigned *pinds,
//         unsigned len_p, unsigned pstart, unsigned pstop)

// valgrind -v --track-origins=yes --leak-check=full ./main
int main(void) { // for valgrinding
   
   unsigned L = 14, n = 30;
   unsigned m = 2<<(L-1);
   double *px = malloc(m*n*sizeof(double));
   double *pPHx = malloc(m*n*sizeof(double));
   unsigned pinds[] = {0, 1, 4, 5};

   srht_rec_had(pPHx, px, m, n, L, pinds, 4, 4, 0);

   free(px);
   free(pPHx);
   
   return 1; 
}
